[CORE-9523] Consumer Group Lag: Set empty shard label

BenPope opened the pull request:
As currently implemented, the metrics contain the shard label, which can add to cardinality over time as the group coordinator can change.

Override the shard label to prevent seastar setting it, to prevent cardinality increase if the group coordinator moves shard.

## Backports Required


- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

* none

IoannisRP commented on src/v/kafka/server/group.cc:
```diff
@@ -3630,7 +3636,7 @@ void group::setup_metrics() {
         }
 
-        if (_enable_group_metrics().consumer_lag) {
+        if (_enable_group_metrics().consumer_lag && need_lag_metrics(_state)) {
```
what is special about the consumer_lag in relation to `group_state`?

BenPope replied:
> what is special about the consumer_lag in relation to `group_state`?

Nothing. I wasn't actually supposed to commit that bit.

BenPope replied:
Moved to a separate commit.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--7e5f8838-eba4-4f6c-9e4e-296f66604573-->
<!--TEST RESULTS-->
<!--7f4e7d86-e7cf-4863-97dd-aa8d00bb1213-->
<!--TEST RESULTS-->
<!--2a40ae7c-14a2-4d8c-b73f-595f37f04188-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/63138">build#63138</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.True | ducktape | https://buildkite.com/redpanda/redpanda/builds/63138#0195950d-d561-4310-848f-7ca77928078e | FLAKY | 1/2 |
| rptest.tests.datalake.datalake_upgrade_test.DatalakeUpgradeTest.test_upload_through_upgrade.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK | ducktape | https://buildkite.com/redpanda/redpanda/builds/63138#01959511-7b71-4eb4-9b2f-433e41a15b79 | FLAKY | 1/2 |
| rptest.tests.node_pool_migration_test.NodePoolMigrationTest.test_migrating_redpanda_nodes_to_new_pool.balancing_mode=off.test_mode=TestMode.TIRED_STORAGE.cleanup_policy=compact.delete | ducktape | https://buildkite.com/redpanda/redpanda/builds/63138#01959511-7b6f-470a-ab23-61e854afd89b | FLAKY | 1/2 |
| rptest.tests.simple_e2e_test.SimpleEndToEndTest.test_relaxed_acks.write_caching=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/63138#01959511-7b6f-4aa7-8fa1-eb2df340833e | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/63218">build#63218</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/63218#0195a3a2-4ddc-46de-a45f-5b2f6a456aac | FLAKY | 1/2 |
| rptest.tests.consumer_group_balancing_test.ConsumerGroupBalancingTest.test_coordinator_nodes_balance | ducktape | https://buildkite.com/redpanda/redpanda/builds/63218#0195a3a2-4ddb-4433-b438-f28230b8d23d | FLAKY | 1/2 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=False.params=.cancellation.dir.in.stage.preparing.use_alias.True | ducktape | https://buildkite.com/redpanda/redpanda/builds/63218#0195a39e-7b35-4524-b67c-f8f073860e00 | FLAKY | 1/2 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.True | ducktape | https://buildkite.com/redpanda/redpanda/builds/63218#0195a39e-7b36-496c-b759-5dc5dd1cb728 | FLAKY | 1/3 |
| rptest.tests.datalake.datalake_upgrade_test.DatalakeUpgradeTest.test_upload_through_upgrade.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK | ducktape | https://buildkite.com/redpanda/redpanda/builds/63218#0195a3a2-4ddc-41c1-b4e8-2b357d899fe0 | FLAKY | 1/3 |
| rptest.tests.datalake.transactions_test.DatalakeTransactionTests.test_with_transactions.cloud_storage_type=CloudStorageType.S3.compaction=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/63218#0195a3a2-4ddb-4d08-90da-4a50a04916c1 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/63365">build#63365</a></summary>


| test_id                                                                                                                                                                             | test_kind   | job_url                                                                                   | test_status   | passed   |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.False | ducktape    | https://buildkite.com/redpanda/redpanda/builds/63365#0195afc4-f776-4204-880a-624b2bd6c3cc | FLAKY         | 1/5      |
| rptest.tests.datalake.custom_partitioning_test.DatalakeCustomPartitioningTest.test_spec_evolution.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP       | ducktape    | https://buildkite.com/redpanda/redpanda/builds/63365#0195afc4-f775-40d1-9b35-74e4204bda2d | FLAKY         | 1/2      |
| rptest.tests.datalake.datalake_e2e_test.DatalakeE2ETests.test_topic_lifecycle.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE                                | ducktape    | https://buildkite.com/redpanda/redpanda/builds/63365#0195afc4-f775-40d1-9b35-74e4204bda2d | FLAKY         | 1/2      |
| rptest.tests.schema_registry_test.SchemaRegistryAutoAuthTest.test_normalize.dataset_type=JSON                                                                                       | ducktape    | https://buildkite.com/redpanda/redpanda/builds/63365#0195afb2-3709-428f-afdc-6117f84041a3 | FLAKY         | 1/5      |
| rptest.tests.schema_registry_test.SchemaRegistryAutoAuthTest.test_normalize.dataset_type=JSON                                                                                       | ducktape    | https://buildkite.com/redpanda/redpanda/builds/63365#0195afc4-f777-4690-a930-0b144eb51da4 | FLAKY         | 1/2      |
| rptest.tests.schema_registry_test.SchemaRegistryTest.test_normalize.dataset_type=JSON                                                                                               | ducktape    | https://buildkite.com/redpanda/redpanda/builds/63365#0195afb2-370b-4af1-ba19-35ffb25fe372 | FLAKY         | 1/3      |
</details>
<!--/body-->

BenPope committed:
Consumer Group Lag: Set shard label to empty

As currently implemented, the metrics contain the shard label,
which can add to cardinality over time as the group coordinator
can change shard.

Reduce cardinality by explicitly setting shard label empty, to
prevent seastar from adding it.

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
Consumer Group Lag: Don't emit metrics if state::dead

A group which has transitioned to dead does not need metrics;
ensure metrics are only emmitted when state is not dead by
checking the state in setup_metrics(), and set_state where
all transitions pass through.

This avoids making requests for partition offsets when not
necessary.

Signed-off-by: Ben Pope <ben@redpanda.com>

StephanDollberg commented:
This is a bit of an anti pattern because of the issue described here: https://github.com/redpanda-data/redpanda/pull/23339

I don't think cardinality itself is an issue anymore as per previous discussions. If you really want to avoid the shard label changing I think best would be to explicitly add the shard label with a static value ("0" or "") as that will prevent seastar from adding it automatically (with the varying shard).

BenPope commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/362e600a69946d0279005f3a2dedac10408d2c72..d64ca1bab52c77b7430b97e77b12c55e56391e58)

* Address review comment, set shard label empty.

IoannisRP reviewed (approved):
lgtm

vbotbuildovich commented:
/backport v25.1.x
